File: BOJ/1753번/1753.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int INF = 300000;
struct info {
    int link;
    int weight;
    bool operator<(const info &a) const { return weight > a.weight; }
};
vector<int> dis;
vector<vector<info > > node;
int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);cout.tie(0);

    int V, E;
    cin >> V >> E;
    node.resize(V + 2);
    dis.assign(V + 2, INF);

    int K;
    cin >> K;
    dis[K] = 0;

    int u, v, w;
    while (E--) {
        cin >> u >> v >> w;
        node[u].push_back({v, w});
    }

    priority_queue<info > q;
    q.push({K, 0});

    while (!q.empty()) {
        auto cur = q.top();
        q.pop();

        if (cur.weight != dis[cur.link]) continue;

        for (auto nxt : node[cur.link]) {
            int distance = dis[cur.link] + nxt.weight;
            if (distance < dis[nxt.link]) {
                dis[nxt.link] = distance;
                q.push({nxt.link, distance});
            }
        }
    }

    for (int i = 1; i <= V; i++) {
        if (dis[i] >= INF) cout << "INF";
        else cout << dis[i];
        cout << "\n";
    }
    return 0;
}

File: BOJ/1753번/1753.kt
import java.util.*
data class Node(val start: Int, val dis: Int): Comparable<Node> {
    override fun compareTo(other: Node) = if (dis < other.dis) -1 else 1
}
fun main() {
    val br = System.`in`.bufferedReader()
    fun input() = br.readLine().split(" ").map(String::toInt)
    val limit = 1e9.toInt()
    val (v, e) = input()
    val k = input()[0] - 1
    val distance = IntArray(v) { limit }
    val link = Array(v) { mutableListOf<Node>() }
    for (it in 0 until e) {
        val (a, b, c) = input()
        link[a - 1].add(Node(b - 1, c))
    }
    val queue = PriorityQueue<Node>()
    distance[k] = 0
    queue.add(Node(k, distance[k]))

    while (queue.isNotEmpty()) {
        val curStart = queue.peek().start
        val curWeight = queue.remove().dis

        if (curWeight != distance[curStart]) continue
        link[curStart].forEach {
            val nextWeight = curWeight + it.dis
            if (nextWeight < distance[it.start]) {
                distance[it.start] = nextWeight
                queue.add(Node(it.start, distance[it.start]))
            }
        }
    }
    for (it in 0 until v) print("${if (distance[it] == limit) "INF" else distance[it]}\n")
}